rasa/shared/core/domain.py

```diff
@@ -309,8 +311,93 @@ def merge_lists_of_dicts(
 
         return self.__class__.from_dict(combined)
 
+    @staticmethod
+    def merge_domain_dicts(
+        domain1: Dict, domain2: Dict, override: bool = False
+    ) -> Dict[Text, Any]:
+        """Merge this domain dict with another one, combining their attributes.
+
+        Used when multiple domain files are configured (in a single directory).
+
+        List attributes like ``intents`` and ``actions`` will be deduped
+        and merged. Single attributes will be taken from `domain1` unless
+        override is `True`, in which case they are taken from `domain2`.
+
+        This is a fix for #10137
+        """
+        if not domain2:  # or domain2.is_empty():
+            return domain1
+
+        if domain1 == {}:  # .is_empty():
+            return domain2
+
+        domain_dict = domain2  # .as_dict(mergeflag=True)
+        combined = domain1  # .as_dict(mergeflag=True)
+
+        def merge_dicts(
+            tempDict1: Dict[Text, Any],
+            tempDict2: Dict[Text, Any],
+            override_existing_values: bool = False,
+        ) -> Dict[Text, Any]:
+
+            if override_existing_values:
+                merge_dicts, b = tempDict1.copy(), tempDict2.copy()
+
+            else:
+                merge_dicts, b = tempDict2.copy(), tempDict1.copy()
+
+            merge_dicts.update(b)
+            return merge_dicts
+
+        def merge_lists(list1: List[Any], list2: List[Any]) -> List[Any]:
+            return sorted(list(set(list1 + list2)))
```

Comment:
There's currently a lot of code duplication. For example the highlighted code is repeated in `merge`. Should we reduce code duplication? Maybe these nested functions can be moved outside of the merge functions and reused?